server/storage/backend/read_tx.go

```diff
@@ -120,6 +130,106 @@ func (baseReadTx *baseReadTx) UnsafeRange(bucketType Bucket, key, endKey []byte,
 	return append(k2, keys...), append(v2, vals...)
 }
 
+func bytesToRev(bytes []byte) int64 {
+	return int64(binary.BigEndian.Uint64(bytes[0:8]))
+}
+
+func (baseReadTx *baseReadTx) UnsafeRangeWithLock(bucketType Bucket, key, endKey []byte, limit int64, b Backend) ([][]byte, [][]byte) {
+	if endKey == nil {
+		// forbid duplicates for single keys
+		limit = 1
+	}
+	if limit <= 0 {
+		limit = math.MaxInt64
+	}
+	if !bucketType.IsSafeRangeBucket() {
+		panic("do not use UnsafeRangeWithLock on non-keys bucket")
+	}
+
+	if endKey == nil {
+		// single key search, key exists either bolt db or read buffer
+		k1, v1 := baseReadTx.unsafeRangeDB(bucketType, key, endKey, limit)
+		if int64(len(k1)) == limit {
+			return k1, v1
+		}
+		k2, v2 := baseReadTx.unsafeRangeBuff(bucketType, key, endKey, limit-int64(len(k1)), b)
+		return append(k2, k1...), append(v2, v1...)
+	}
```

Comment:
why do we want to deal with a single key case separately?

Reply:
Because when we range with limit, we need search data from newest to oldest. Data from unsafeRangeDB is older than unsafeRangeBuff, which is wrong order for batch range. But for single key, data exists either DB or buff, so order is no needed.